Kernel/VirtIO/VirtIOConsole.cpp

```diff
@@ -73,115 +60,135 @@ bool VirtIOConsole::handle_device_config_change()
+                process_control_message(*message);
+            });
 
-        VirtIOQueueChain new_chain(queue);
-        if (remaining_space != 0) {
-            new_chain.add_buffer_to_chain(used_space.offset(used), remaining_space, BufferType::DeviceWritable);
+            supply_chain_and_notify(CONTROL_RECEIVEQ, popped_chain);
+            popped_chain = queue.pop_used_buffer_chain(used);
         }
-        supply_chain_and_notify(RECEIVEQ, new_chain);
-
-        evaluate_block_conditions();
-        break;
-    }
-    case TRANSMITQ: {
-        ScopedSpinLock ringbuffer_lock(m_transmit_buffer->lock());
-        auto& queue = get_queue(TRANSMITQ);
+    } else if (queue_index == CONTROL_TRANSMITQ) {
+        ScopedSpinLock ringbuffer_lock(m_control_transmit_buffer->lock());
+        auto& queue = get_queue(CONTROL_TRANSMITQ);
         ScopedSpinLock queue_lock(queue.lock());
         size_t used;
         VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);
         do {
             popped_chain.for_each([this](PhysicalAddress address, size_t length) {
-                m_transmit_buffer->reclaim_space(address, length);
+                m_control_transmit_buffer->reclaim_space(address, length);
             });
             popped_chain.release_buffer_slots_to_queue();
             popped_chain = queue.pop_used_buffer_chain(used);
         } while (!popped_chain.is_empty());
-        // Unblock any IO tasks that were blocked because can_write() returned false
-        evaluate_block_conditions();
-        break;
+    } else {
+        auto port_index = queue_index / 2 == 0 ? 0 : (queue_index - 2) / 2;
+        m_ports.get(port_index).value()->handle_queue_update({}, queue_index);
     }
-    default:
-        VERIFY_NOT_REACHED();
-    }
-}
-
-bool VirtIOConsole::can_read(const FileDescription&, size_t) const
-{
-    return m_receive_buffer->used_bytes() > 0;
 }
 
-KResultOr<size_t> VirtIOConsole::read(FileDescription& desc, u64, UserOrKernelBuffer& buffer, size_t size)
+void VirtIOConsole::setup_multiport()
 {
-    if (!size)
-        return 0;
+    m_control_receive_buffer = make<RingBuffer>("VirtIOConsole control receive queue", CONTROL_BUFFER_SIZE);
+    m_control_transmit_buffer = make<RingBuffer>("VirtIOConsole control receive queue", CONTROL_BUFFER_SIZE);
 
-    if (!can_read(desc, size))
-        return EAGAIN;
-
-    ScopedSpinLock ringbuffer_lock(m_receive_buffer->lock());
-    auto& queue = get_queue(RECEIVEQ);
+    auto& queue = get_queue(CONTROL_RECEIVEQ);
     ScopedSpinLock queue_lock(queue.lock());
+    VirtIOQueueChain chain(queue);
+    auto offset = 0ul;
+    auto buffer_size = 8;
 
-    auto bytes_copied = m_receive_buffer->copy_data_out(size, buffer);
-    m_receive_buffer->reclaim_space(m_receive_buffer->start_of_used(), bytes_copied.value());
-    evaluate_block_conditions();
+    while (offset < CONTROL_BUFFER_SIZE) {
```

Comment:
we should probably `VERIFY` (or even static_assert if we go with constexpr-static route) that buffer_size divides CONTROL_BUFFER_SIZE